drivers: gpio: xlnx_axi: Add MMU support

drakargx opened the pull request:
The current gpio_xlnx_axi will cause a data abort exception on SoCs that have a memory protection unit enabled. This patch modifies the driver to use the MMIO helper macros to add virtual address mapping for those SoCs.

Commit:
drivers: gpio: xlnx_axi: add mmu support
Adds MMU support to the Xilinx AXI GPIO driver, allowing it to be used on SoCs with an enabled memory protection unit, such as the Zynq-7000.

Signed-off-by: Nigel Hearn <nhearn@tuta.io>

Commit:
run uncrustify on gpio_xlnx_axi.c

Signed-off-by: Nigel Hearn <nhearn@mmm.com>

drakargx commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -229,6 +222,8 @@ static int gpio_xlnx_axi_init(const struct device *dev)
 {
 	struct gpio_xlnx_axi_data *data = dev->data;
 
+	DEVICE_MMIO_MAP(dev, K_MEM_CACHE_NONE);
```
My understanding is that if dual-channel GPIO is enabled, then the driver will create two devices that will be initialized separately. Will having both devices calling DEVICE_MMIO_MAP cause problems? The second channel will reside in the first's virtual address mapping, but will attempt to call DEVICE_MMIO_MAP with a different size and starting address.

henrikbrixandersen replied:
I am not sure mapping the same region twice is allowed.

drakargx replied:
If that is the case, would something like this be allowed:

```
#ifdef CONFIG_MMU
static int gpio_xlnx_axi_mmio_init(const struct device *dev)
{
	const struct gpio_xlnx_axi_config *config = dev->config;
	struct gpio_xlnx_axi_data *data = dev->data;

	device_map(&data->mmio, config->base, config->size, K_MEM_CACHE_NONE);

	gpio_xlnx_axi_write_data(dev, data->dout);
	gpio_xlnx_axi_write_tri(dev, data->tri);

	return 0;
}
#endif

static int gpio_xlnx_axi_init(const struct device *dev)
{
	struct gpio_xlnx_axi_data *data = dev->data;

#ifdef CONFIG_MMU
	const device_handle_t *handles;
	size_t handle_count;
	const struct device *parent_device;
	struct gpio_xlnx_axi_data *parent_data;

	handles = device_required_handles_get(dev, &handle_count);

	parent_device = device_from_handle(handles[0]);
	parent_data = parent_device->data;
	data->mmio = parent_data->mmio + GPIO2_DATA_OFFSET;
#endif

	gpio_xlnx_axi_write_data(dev, data->dout);
	gpio_xlnx_axi_write_tri(dev, data->tri);

	return 0;
}
```

The gpio2 nodes would be moved down to `PRE_KERNEL_2` priority as they currently are initialized before their parent node. If `CONFIG_MMU` is defined then the parent gpio node will initialize using `gpio_xlnx_axi_mmio_init()` and will create the memory-mapped region. The gpio2 nodes would initialize using `gpio_xlnx_axi_init()` and would look for their parent to get the base address to write to. 

If `CONFIG_MMU` is not defined then both the gpio and gpio2 nodes will initialize with gpio_xlnx_axi_init.

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -247,76 +242,111 @@ static const struct gpio_driver_api gpio_xlnx_axi_driver_api = {
+		DT_NODE_HAS_COMPAT(DT_CHILD(DT_DRV_INST(n), gpio2), \
+				   xlnx_xps_gpio_1_00_a_gpio2),	    \
+		DT_NODE_HAS_STATUS(DT_CHILD(DT_DRV_INST(n), gpio2), \
+				   okay)			    \
 		)
```
Please avoid unrelated whitespace changes. It makes the PR much harder to review.

drakargx replied:
The whitespace change was the result of running uncrustify on the file before commiting.

henrikbrixandersen commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -247,76 +242,111 @@ static const struct gpio_driver_api gpio_xlnx_axi_driver_api = {
+	static const struct gpio_xlnx_axi_config			      \
+			    gpio_xlnx_axi_##n##_2_config = {		      \
+		._mmio = {						      \
+			.phys_addr = DT_INST_REG_ADDR(n) + GPIO2_DATA_OFFSET, \
+			.size = DT_INST_REG_SIZE			      \
+		},							      \
```
This is clearly wrong. You are adding an offset, but using the size without taking the offset into consideration.

drakargx commented on drivers/gpio/gpio_xlnx_axi.c:
```diff
@@ -32,15 +32,14 @@
 
 struct gpio_xlnx_axi_config {
-	/* gpio_driver_config needs to be first */
+	DEVICE_MMIO_ROM;
 	struct gpio_driver_config common;
```
`DEVICE_MMIO_ROM` and `DEVICE_MMIO_RAM` also have a requirement that they need to be the first member of the struct, so perhaps it would have been better to not rely on the use of those macros and to add address, size and mmio address information to the struct (omitting mmio address and size if `CONFIG_MMU` is not defined.

drakargx committed:
Modified gpio_xlnx_axi driver to call device_map() once per node

Updates the gpio_xlnx_axi driver to call device_map() once per devicetree node with a "xlnx,xps-gpio-1.00.a" compatible field.

This is done by adding fields to gpio_xlnx_axi_config and gpio_xlnx_axi_data if CONFIG_MMU is defined. The xlnx,xps-gpio-1.00.a compatible node's data struct now holds a pointer to the gpio2 device, if it exists. This node will be the only one making the call to device_map().

Signed-off-by: Nigel Hearn <nhearn@tuta.io>
